Implement equality operator for attributeset class

phate opened the pull request:
This PR reimplements the internals of the attributeset class such that we can properly support `operator==()`.

Closes #525

phate commented on jlm/llvm/ir/attribute.hpp:
```diff
@@ -267,181 +220,148 @@ class int_attribute final : public enum_attribute
-      FIXME: Ah, since this is not a real set, we cannot cheaply implement a comparison.
-    */
-    return false;
+    // FIXME: I have no better idea of how to handle this currently
+    return std::hash<ptrdiff_t>()(reinterpret_cast<ptrdiff_t>(&attribute));
```
@haved @caleridas Any better idea?

haved replied:
This might be problematic as two `type_attribute` can compare equal with `operator==` yet have different hashes.

Would hashing the kind and type and xoring them together work?

haved replied:
@phate I checked the definition of std::hash, and they require == equality to imply identical hashes

phate replied:
Yeah, this won't work properly unless we implement hashing capabilities for `rvsdg::type`. I will take a stab at it in a new PR, then we can merge this one.

phate replied:
Fixed.

phate commented:
FYI @ProgrammingLouis

phate commented:
@caleridas @haved I finally found some time to fix this. The changes since last review:
1. Fixed hashing of types (and improved the combining of the other hashes)
2. Added a unit test for attribute set (in-)equality

FYI @ProgrammingLouis

phate commented:
@haved @caleridas Anyone, please?

haved reviewed (dismissed):
Looks good, I only have a question about one of the casts.

Also in your PR description you say `operator=()`, I assume you mean `operator==()`

phate commented on jlm/llvm/backend/jlm2llvm/jlm2llvm.cpp:
```diff
@@ -271,42 +271,54 @@ convert_attribute_kind(const attribute::kind & kind)
-  {
-    auto & llvmctx = ctx.llvm_module().getContext();
+  auto & llvmContext = ctx.llvm_module().getContext();
+  auto ea = dynamic_cast<const enum_attribute *>(&attribute);
```
Ah, that is nonsense. I will fix it.

phate replied:
Removed.

phate commented:
@haved I removed the `dynamic_cast<>()` and fixed the description. I need a new review though.
